compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -383,6 +403,106 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
+            (Some(alias_lhs), Some(alias_rhs)) => {
+                // Evaluate all 3 potential candidats for the alias' being equal
+                let lhs_to_rhs = evaluate_normalizes_to(self, alias_lhs, goal.predicate.1);
+                let rhs_to_lhs = evaluate_normalizes_to(self, alias_rhs, goal.predicate.0);
+                let opt_substs_eq = match alias_lhs.def_id == alias_rhs.def_id {
+                    true => Some(self.infcx.probe(|_| {
+                        let nested_goals = self.infcx.eq(goal.param_env, alias_lhs, alias_rhs)?;
+                        self.evaluate_all_and_make_canonical_response(nested_goals)
+                    })),
+                    false => None,
+                };
+
+                if let Some(substs_eq) = opt_substs_eq {
+                    // If we have `alias_eq(<T as Trait>::Assoc, <T as Trait>::Assoc)` we will end up here
+                    if lhs_to_rhs.is_err() && rhs_to_lhs.is_err() {
+                        return substs_eq;
+                    }
+
+                    // If we end up with `alias_eq(<?0 as Trait>::Assoc, <T as Trait>::Assoc)` we cant
+                    // unify `?0 = T` during coherence as there may exist some other concrete type for `?0`
+                    // that would normalize to `<T as Trait>::Assoc`.
+                    //
+                    // We don't want to return `substs_eq` in the case of it being `NoSolution` as this
+                    // goal may still be solvable by one of the `normalizes_to` goals.
+                    if (lhs_to_rhs.is_err() || rhs_to_lhs.is_err())
+                        && !self.infcx.intercrate
+                        && substs_eq.is_ok()
+                    {
+                        return substs_eq;
+                    }
+
+                    // If we relating the substs of the two aliases succeeds and introduces no constraints that
+                    // means we have something like `alias_eq(<?0 as Trait>::Assoc, <?0 as Trait>::Assoc)` and
+                    // should return a response of `Ok(Yes)` with no constraints regardless of what the `normalizes_to`
+                    // goals evaluated to.
+                    if let Ok(response) = &substs_eq {
+                        if response.value.certainty == Certainty::Yes
+                            && !response.has_inference_or_external_constraints()
+                        {
+                            return substs_eq;
+                        }
+                    }
+                }
+
+                match (lhs_to_rhs, rhs_to_lhs) {
+                    (Err(NoSolution), Err(NoSolution)) => Err(NoSolution),
+                    (Ok(response), Err(NoSolution)) | (Err(NoSolution), Ok(response)) => {
+                        Ok(response)
+                    }
+                    (Ok(response_lhs), Ok(response_rhs)) => {
+                        if response_lhs == response_rhs {
```

Comment:
```rust
if substs_eq.map_or(false, |response| response.value.certainty == Certainty::Yes
                            && !response.has_inference_or_external_constraints()) {
    substs_eq   
}
```

Reply:
personally find this kinda harder to read than just the `if let` + `if` that exists currently, too bad let chains dont workotherwise it'd be a bit nicer.

Reply:
are you thinking that this condition will never be true for the `(Ok(_), Err(NoSolution))` arm? :thinking:

Reply:
Ah I guess before we removed the "when not in intercrate, relate substs" we would have been returning `substs_eq` anyway in that case, but now that its removed thats no longer the case